Sources/StreamChat/Repositories/AuthenticationRepository.swift

```diff
@@ -0,0 +1,190 @@
+        clientUpdater: ChatClientUpdater,
+        tokenExpirationRetryStrategy: RetryStrategy,
+        timerType: Timer.Type
+    ) {
+        self.apiClient = apiClient
+        self.databaseContainer = databaseContainer
+        self.clientUpdater = clientUpdater
+        self.tokenExpirationRetryStrategy = tokenExpirationRetryStrategy
+        self.timerType = timerType
+
+        fetchCurrentUser()
+    }
+
+    private func fetchCurrentUser() {
+        var currentUserId: UserId?
+
+        let context = databaseContainer.viewContext
+        if Thread.isMainThread {
+            currentUserId = context.currentUser?.user.id
+        } else {
+            context.performAndWait {
+                currentUserId = context.currentUser?.user.id
+            }
+        }
+        self.currentUserId = currentUserId
+    }
+
+    /// Sets the user token. This method is only needed to perform API calls without connecting as a user.
+    /// You should only use this in special cases like a notification service or other background process
+    func setToken(token: Token) {
+        currentToken = token
+    }
+
+    /// Establishes a connection for a  user.
+    /// - Parameters:
+    ///   - userInfo:       The user information that will be created OR updated if it exists.
+    ///   - tokenProvider:  The block to be used to get a token.
+    func connectUser(with userInfo: UserInfo?, tokenProvider: @escaping TokenProvider, completion: @escaping (Error?) -> Void) {
+        connect(userInfo: userInfo, tokenProvider: tokenProvider, completion: completion)
+    }
+
+    /// Establishes a connection for a guest user.
+    /// - Parameters:
+    ///   - userInfo: The user information that will be created OR updated if it exists.
+    func connectGuestUser(userInfo: UserInfo, completion: @escaping (Error?) -> Void) {
+        connect(
+            userInfo: userInfo,
+            tokenProvider: { [weak self] completion in
+                self?.fetchGuestToken(userInfo: userInfo, completion: completion)
+            },
+            completion: completion
+        )
+    }
+
+    private func connect(userInfo: UserInfo?, tokenProvider: @escaping TokenProvider, completion: @escaping (Error?) -> Void) {
+        self.tokenProvider = tokenProvider
+        getToken(userInfo: userInfo, tokenProvider: tokenProvider, completion: completion)
+    }
+
+    func logOutUser() {
+        log.debug("Logging out user", subsystems: .authentication)
+        tokenProvider = nil
+    }
+
+    func refreshToken(completion: @escaping (Error?) -> Void) {
+        guard let tokenProvider = tokenProvider else {
+            let error = ClientError.MissingTokenProvider()
+            log.assertionFailure(error.localizedDescription)
+            completion(error)
+            return
+        }
+
+        let tokenProviderCheckingSuccess: TokenProvider = { [weak self] completion in
+            tokenProvider { result in
+                if case .success = result {
```

Comment:
Shouldn't we handle the error case here as well?

Reply:
Not sure I get your point here. If the token refresh fails, we just return the error, and let it up to the callee to decide what to do.
It is only in the case of a success that we reset the retry strategy

Reply:
thanks for clarifying, my bad - somehow I've seen the completion inside the if-case 😁